GENOME = "yeast.fa"
INDEX = "yeast"
N = [1,2,3,4,5,6,7,8,9,10]


"""
rule all:
    input:
	"sample.trimmed.fq.gz"
"""
rule ribofilio: 
    input:
        fp = "{SAMPLE1}.bed",
	rna = "{SAMPLE2}.bed",
        fake = expand("{{SAMPLE1}}_{IDX}.bed", IDX =[1,2,3] ),
	genome = GENOME 
    output: 
         "{SAMPLE1}_{SAMPLE2}.regression.log"
    params:
        prefix="{SAMPLE1}_{SAMPLE2}"

    shell: "python ../src/ribofilio.py -t {input.genome} -f {input.fp} -r {input.rna} -o {params.prefix} && python ../src/ribofilio.py -t {input.genome} -f {input.fake} -r {input.rna} -o {params.prefix} -p 0 "
  
rule run_subset:
    input:
        fp = "{SAMPLE1}.bed",
        rna = "{SAMPLE2}.bed",
	smallset ="{SUBSET}.txt",
        genome = GENOME
    output:
         "{SAMPLE1}_{SAMPLE2}_{SUBSET}.regression.log"
    params:
        prefix="{SAMPLE1}_{SAMPLE2}_{SUBSET}"

    shell: "python ../src/ribofilio.py -t {input.genome} -f {input.fp} -r {input.rna} -s {input.smallset} -o {params.prefix}"
 
rule  shuffle: 
    input: 
        fp = "{SAMPLE}.bed",
        genome = GENOME
    params: 
        prefix = "{IDX}" 
    output:
        "{SAMPLE}_{IDX}.bed" 
    shell: 
        "python src/shuffle.py -t {input.genome} -i {params.prefix} -b {input.fp} > {output}"

          
rule tobed: 
	input: 
             "{SAMPLE}.bowtie2.bam" 
	output: 
	     "{SAMPLE}.bed"
	shell: 
            "bedtools bamtobed -i {input} > {output}"

rule index: 
     input: 
        genome = GENOME
     output: 
        INDEX
     shell: 
         "bowtie2-build {input.genome} {output} && touch {output}"

rule tobam:
      input:
          index  = INDEX,
          sample = "galore/{READS}_trimmed.fq.gz"
      output:
          "{READS}.bowtie2.bam"  
      shell: 
            "bowtie2 -x {input.index} -U {input.sample} -S | samtools view -S -b > {output}" 

