configfile: "config.yaml"
N =  range(1, config['N_Shuffled'])

rule all:
    input:
        expand("{FP}_{RNA}.{group}.slopes.png",FP = config['FP'], RNA = config['RNA'], group=config['GO']),
        expand("{FP}_{RNA}.{group}.slopes.png",FP = config['FP'], RNA = config['RNA'], group=config['GL']),
        expand("{FP}_{RNA}.slopes.png",FP = config['FP'], RNA = config['RNA']) 

rule ribofilio: 
    input:
         fp = "{SAMPLE1}.bed",
	 rna = "{SAMPLE2}.bed",
	 genome = config['GENOME'] 
    output: 
         "{SAMPLE1}_{SAMPLE2}.regression.log"
    params:
        prefix="{SAMPLE1}_{SAMPLE2}"
    shell:
        "python ../src/ribofilio.py -t {input.genome} -f {input.fp} -r {input.rna} -o {params.prefix} "

rule ribo_all: 
    input:
       fp = "{FP}_{RNA}.regression.log",
       shuffled = expand("{{FP}}.{IDX}.bed", IDX=N),
       dummies = expand("{{FP}}.{IDX}_{{RNA}}.regression.log", IDX= N)
    params:
       all = "{FP}_{RNA}.all" 
    output:
       "{FP}_{RNA}.slopes.png"
    shell: 
       """
       cat {input.fp} {input.dummies} > {params.all}
       python src/stats.py {params.all}
       """    
       
rule subset:
    input:
        fp = "{SAMPLE1}.bed",
        rna = "{SAMPLE2}.bed",
	subset = "{SUBSET}.txt",
        genome = config['GENOME']
    output:
        "{SAMPLE1}_{SAMPLE2}-{SUBSET}.regression.log"
    params:
        prefix="{SAMPLE1}_{SAMPLE2}-{SUBSET}"
    shell: 
        "python ../src/ribofilio.py -t {input.genome} -f {input.fp} -r {input.rna} -s {input.subset} -o {params.prefix}"

rule subset_all: 
    input:
       fp = "{FP}_{RNA}-{SUBSET}.regression.log",
       go = expand("{GO}.txt", GO= config['GO']), 
       shuffled =  expand("{{FP}}.{IDX}.bed", IDX=N),
       dummies = expand("{{FP}}.{IDX}_{{RNA}}-{{SUBSET}}.regression.log", IDX= N)
    params:
       all = "{FP}_{RNA}.{SUBSET}.all"
    output:
       "{FP}_{RNA}.{SUBSET}.slopes.png"
    shell:
       """
       cat {input.fp} {input.dummies} > {params.all}
       python src/stats.py {params.all}
       """

rule get_GO:
     output:
        expand("{GO}.txt", GO=config['GO'])
     params: 
        go = expand("{GO}", GO=config['GO'] )
     shell:
        """
         Rscript src/getGO.R {params.go}
        """

rule cluster: 
     output: 
         "yeast02000.txt",
         "yeast20004000.txt",
         "yeast40006000.txt",
         "yeast6000100000.txt"
     shell: 
          """
          python src/gcluster.py yeast.fa yeast02000.txt --max 2000 --min 0
          python src/gcluster.py yeast.fa yeast20004000.txt --max 4000 --min 2000
          python src/gcluster.py yeast.fa yeast40006000.txt --max 6000 --min 4000
          python src/gcluster.py yeast.fa yeast6000100000.txt --max 100000 --min 6000
          """ 

rule shuffle:
    input:
        fp = "{SAMPLE}.bed",
        genome = config['GENOME']
    output:
        expand("{{SAMPLE}}.{IDX}.bed", IDX=N)
    params: 
        config['N_Shuffled']
    shell:
         """
         python src/shuffle.py -t {input.genome} -b {input.fp} -o {wildcards.SAMPLE} -n {params}
         """ 

rule index: 
     input: 
        genome = config['GENOME']
     output: 
        config['INDEX']
     shell: 
        "bowtie2-build {input.genome} {output} && touch {output}"

rule trim: 
    input:
        sample = "{SAMPLE}.fastq",
	index = config['INDEX']
    output:
        "galore/{SAMPLE}_trimmed.fq.gz", 
    params: 
       adapter =config['ADAPTERS']
    shell:
        """
        trim_galore --gzip --retain_unpaired --trim1 -a {params.adapter} --fastqc --fastqc_args "--outdir fastqc" -o galore {input.sample}   
        """ 

rule tobam:
      input:
          index  = config['INDEX'],
          sample = "galore/{READS}_trimmed.fq.gz"
      output:
          "{READS}.sam",
          "{READS}.bowtie2.bam"
      shell: 
          """ 
           bowtie2 -x {input.index} -U {input.sample} -S {output[0]}
           samtools view -S -b {output[0]} > {output[1]}
          """ 	

rule tobed: 
        input: 
             "{SAMPLE}.bowtie2.bam", 
             genome = config['GENOME'] 
        output: 
	     "{SAMPLE}.bed" 
	run: 
            """
            bedtools bamtobed -i {input} > {output}
            """


rule getGenome:
     output:
           config['GENOME']
     shell:
        """
        curl -O ftp://ftp.ensembl.org/pub/release-98/fasta/saccharomyces_cerevisiae/cdna/Saccharomyces_cerevisiae.R64-1-1.cdna.all.fa.gz
        gunzip Saccharomyces_cerevisiae.R64-1-1.cdna.all.fa.gz
        mv  Saccharomyces_cerevisiae.R64-1-1.cdna.all.fa {output}
        """
