configfile: "config.yaml"

#
# run with --use-conda 
#

rule all:
    input:
        expand("{GO}.txt", GO=config['GO'] ),     
        expand("{FP}_{RNA}.regression.log", FP = config['D1_FP'], RNA = config['D1_RNA']),
        expand("{FP}_{RNA}-{SUBSET}.regression.log", FP = config['D1_FP'], RNA = config['D1_RNA'], SUBSET=config['GL']), 
        expand("{FP}_{RNA}-{SUBSET}.regression.log", FP = config['D1_FP'], RNA = config['D1_RNA'], SUBSET=config['GO']),
        expand("{FP}_{RNA}.regression.log", FP = config['D2_FP'], RNA = config['D2_RNA']),
        expand("{FP}_{RNA}-{SUBSET}.regression.log", FP = config['D2_FP'], RNA = config['D2_RNA'], SUBSET=config['GL']),
        expand("{FP}_{RNA}-{SUBSET}.regression.log", FP = config['D2_FP'], RNA = config['D2_RNA'], SUBSET=config['GO']),
        expand("{FP}_{RNA}.regression.log", FP = config['D3_FP'], RNA = config['D3_RNA']),
        expand("{FP}_{RNA}.regression.log", FP = config['D4_FP'], RNA = config['D4_RNA']),
        expand("{FP}_{RNA}-{SUBSET}.ttest.log", FP = config['D1_FP'], RNA = config['D1_RNA'], SUBSET=config['GO']),
        expand("{FP}_{RNA}-{SUBSET}.ttest.log", FP = config['D1_FP'], RNA = config['D1_RNA'], SUBSET=config['GL']),
        expand("{FP}_{RNA}-{SUBSET}.ttest.log", FP = config['D2_FP'], RNA = config['D2_RNA'], SUBSET=config['GO']),
        expand("{FP}_{RNA}-{SUBSET}.ttest.log", FP = config['D2_FP'], RNA = config['D2_RNA'], SUBSET=config['GL']),
        expand("{FP1}_{RNA1}.{FP2}.{RNA2}.ttest.log", FP1 = config['D1_FP'], RNA1 = config['D1_RNA'], FP2 = config['D2_FP'], RNA2 = config['D2_RNA']) 
rule compare: 
   input: 
        "{SAMPLE1}_{SAMPLE2}.regression.log",
        "{SAMPLE3}_{SAMPLE4}.regression.log", 
   output: 
        "{SAMPLE1}_{SAMPLE2}.{SAMPLE3}.{SAMPLE4}.ttest.log" 
   conda: 'env/env-ttest.yaml'
   shell: 
     """
       python scripts/ttest.py {input[0]} {input[1]} > {output}
     """

rule compare_subset:
   input: 
       "{SAMPLE1}_{SAMPLE2}.regression.log",
       "{SAMPLE1}_{SAMPLE2}-{SUBSET}.regression.log"
   output: 
       "{SAMPLE1}_{SAMPLE2}-{SUBSET}.ttest.log"
   conda: 'env/env-ttest.yaml'
   shell: 
       """
       python scripts/ttest.py {input[0]} {input[1]} > {output}
       """
     
rule ribofilio: 
    input:
         fp = "{SAMPLE1}.bed",
	 rna = "{SAMPLE2}.bed",
	 genome = config['GENOME'] 
    output: 
         "{SAMPLE1}_{SAMPLE2}.regression.log"
    params:
        prefix="{SAMPLE1}_{SAMPLE2}", 
        binsize = config['BINSIZE']
    conda: 'env/env-ribo.yaml'
    shell:
        "python {config[src]}/ribofilio.py -t {input.genome} -f {input.fp} -r {input.rna} -b {params.binsize} -o {params.prefix} "

       
rule subset:
    input:
        fp = "{SAMPLE1}.bed",
        rna = "{SAMPLE2}.bed",
	subset = "{SUBSET}.txt",
        genome = config['GENOME']
    output:
        "{SAMPLE1}_{SAMPLE2}-{SUBSET}.regression.log"
    params:
        prefix="{SAMPLE1}_{SAMPLE2}-{SUBSET}",
        binsize = config['BINSIZE']
    conda: 'env/env-ribo.yaml'
    shell: 
        "python {config[src]}/ribofilio.py -t {input.genome} -f {input.fp} -r {input.rna} -s {input.subset} -b {params.binsize} -o {params.prefix}"


rule get_GO:
     output:
        expand("{GO}.txt", GO=config['GO'])
     params: 
        go = expand("{GO}", GO=config['GO'] )
     conda: 'env/env-go.yaml'
     shell:
         """
         Rscript {config[scripts]}/getGO.R {params.go}
         """

rule cluster: 
     output: 
        expand("{GL}.txt", GL=config['GL'])
     conda: 'env/env-cluster.yaml'   
     shell: 
          """
          python {config[scripts]}/gcluster.py yeast.fa yeast0500.txt -n 0 -x 500
	  python {config[scripts]}/gcluster.py yeast.fa yeast5001000.txt -n 500 -x 1000
          python {config[scripts]}/gcluster.py yeast.fa yeast01000.txt -n 0 -x 1000
	  python {config[scripts]}/gcluster.py yeast.fa yeast10002000.txt -n 1000 -x 2000
	  python {config[scripts]}/gcluster.py yeast.fa yeast20003000.txt -n 2000 -x 3000
	  python {config[scripts]}/gcluster.py yeast.fa yeast30004000.txt -n 3000 -x 4000
	  python {config[scripts]}/gcluster.py yeast.fa yeast40005000.txt -n 4000 -x 5000
	  python {config[scripts]}/gcluster.py yeast.fa yeastg5000.txt -n 5000
          """ 


rule index: 
     input: 
        genome = config['GENOME']
     output: 
        config['INDEX']
     conda: 'env/env-index.yaml'
     shell: 
        "bowtie2-build {input.genome} {output} && touch {output}"

rule trim: 
    input:
        sample = "{SAMPLE}.fastq",
	index = config['INDEX']
    output:
        "galore/{SAMPLE}_trimmed.fq.gz", 
    params: 
       adapter =config['ADAPTERS']
    conda: 'env/env-trim.yaml'
    shell:
        """
        trim_galore --gzip --retain_unpaired --trim1 -a {params.adapter} --fastqc --fastqc_args "--outdir fastqc" -o galore {input.sample}   
        """ 

rule tobam:
      input:
          index  = config['INDEX'],
          sample = "galore/{READS}_trimmed.fq.gz"
      output:
          "{READS}.sam",
          "{READS}.bowtie2.bam"
      conda: 'env/env-align.yaml'
      shell: 
          """ 
           bowtie2 -x {input.index} -U {input.sample} -S {output[0]}
           samtools view -S -b {output[0]} > {output[1]}
          """ 	

rule tobed: 
        input: 
             "{SAMPLE}.bowtie2.bam", 
             genome = config['GENOME'] 
        output: 
	     "{SAMPLE}.bed"
        conda: 'env/env-format.yaml' 
	shell: 
            """
            bedtools bamtobed -i {input} > {output}
            """


rule getGenome:
     output:
           config['GENOME']
     shell:
        """
        curl -O ftp://ftp.ensembl.org/pub/release-98/fasta/saccharomyces_cerevisiae/cdna/Saccharomyces_cerevisiae.R64-1-1.cdna.all.fa.gz
        gunzip Saccharomyces_cerevisiae.R64-1-1.cdna.all.fa.gz
        mv  Saccharomyces_cerevisiae.R64-1-1.cdna.all.fa {output}
        """
